File: L2ScriptMaker/CheckUpdate.vb
Public Class CheckUpdate

    Private Sub StartButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles StartButton.Click

        Dim sUrlBase As String = "http://dragon.ur.ru/l2sm/"
        'Dim sUrlBase As String = "http://server/l2sm/"
        Dim l2tempver As String = System.IO.Path.GetTempFileName
        Dim sVersion As String
        Dim sUrlFile As String

        If GetFileFromHttp(sUrlBase & "l2scriptmaker.txt", l2tempver) = False Then
            MessageBox.Show("Can't download version file from site.", "Can't read version", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Exit Sub
        End If


        Dim VerRead As New System.IO.StreamReader(l2tempver, System.Text.Encoding.Default, True, 1)
        sVersion = VerRead.ReadLine
        sUrlFile = VerRead.ReadLine
        CheckStatLabel.Text = "Version detected: " & sVersion
        CheckStatLinkLabel.Text = VerRead.ReadLine

        VerRead.Close()
        System.IO.File.Delete(l2tempver)

        If Application.ProductVersion = sVersion Then
            If MessageBox.Show("You have lastest release" & vbNewLine & "Want to download again?", "No new", MessageBoxButtons.YesNo, MessageBoxIcon.Information) = Windows.Forms.DialogResult.No Then
                Exit Sub
            End If
        Else
            If MessageBox.Show("Found new build! " & sVersion & vbNewLine & "Want to download new?", "Found update", MessageBoxButtons.YesNo, MessageBoxIcon.Information) = Windows.Forms.DialogResult.No Then
                Exit Sub
            End If

        End If

        If GetFileFromHttp(sUrlBase & sUrlFile, sUrlFile) = False Then
            MessageBox.Show("Can't download file from site.", "Can't download", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Exit Sub
        End If

        MessageBox.Show("Download new update success. Last build is " & sVersion & vbNewLine & "Note: close tool, extract new build from file " & sUrlFile & " in your tool folder", "Success.", MessageBoxButtons.OK, MessageBoxIcon.Information)

    End Sub

    Private Sub QuitButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles QuitButton.Click
        Me.Dispose()
    End Sub

    Private Function GetFileFromHttp(ByVal FileUrl As String, ByVal FileDisk As String) As Boolean
        Dim fs As New System.Net.WebClient

        fs.Encoding = System.Text.Encoding.GetEncoding(1250)

        Dim CachePol As New System.Net.Cache.RequestCachePolicy(System.Net.Cache.RequestCacheLevel.NoCacheNoStore)
        fs.CachePolicy = CachePol

        Try
            fs.DownloadFile(FileUrl, FileDisk)
        Catch ex As Exception
            If ex.Message = "The remote server returned an error: (404) Not Found." Then
                'MessageBox.Show("Error on Lineage II Update system. Report to support.")
                Return False
            End If
            'UpdStatus.Text = ex.Message
            MessageBox.Show(ex.Message)
            Return False
        End Try


        Return True
    End Function

    Private Sub SendMailButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles SendMailButton.Click
        Try
            Dim Feedback As New System.Net.Mail.MailMessage(MailFromTextBox.Text, "hellfire@reborn.ru", SubjectTextBox.Text, TextBox.Text)
            Dim Mailclient As New System.Net.Mail.SmtpClient(MailSrvTextBox.Text)
            Mailclient.UseDefaultCredentials = True
            Mailclient.Send(Feedback)
            MessageBox.Show("Send complete.", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information)
        Catch ex As FormatException
            MessageBox.Show(ex.Message)
        Catch ex As System.Net.Mail.SmtpException
            MessageBox.Show(ex.Message)
        End Try

    End Sub


    Private Sub CheckStatLinkLabel_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles CheckStatLinkLabel.Click
        'Process.Start("iexplore.exe " & CheckStatLinkLabel.Text)
        Process.Start("iexplore.exe", CheckStatLinkLabel.Text)
    End Sub
End Class

File: L2ScriptMaker/CheckUpdate.cs
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using L2ScriptMaker.Core.Logger;

namespace L2ScriptMaker
{
	public partial class CheckUpdate : Form
	{
		private const string SupportEmail = "DmitryVHF@outlook.com";
		private const string BaseUrl = "http://dragon.ur.ru/l2sm/";

		private readonly ILogger _logger;

		public CheckUpdate()
		{
			InitializeComponent();

			_logger = new Logger(nameof(CheckUpdate));
		}

		private async void StartButton_Click(object sender, EventArgs e)
		{
			string l2Tempver = System.IO.Path.GetTempFileName();
			string sVersion;
			string sFile;
			string sUrlFile;

			_logger.Write(LogLevel.Information, "Check available update from [" + BaseUrl + "]");
			_logger.Write(LogLevel.Information, "Current version [" + Application.ProductVersion + "]");

			if (await GetFileFromHttp(BaseUrl + "l2scriptmaker.txt", l2Tempver) == false)
			{
				_logger.Write(LogLevel.Error, "Check file not exists [l2scriptmaker.txt]");
				MessageBox.Show("Can't download version file from site.", "Can't read version", MessageBoxButtons.OK,
					MessageBoxIcon.Error);
				return;
			}
			
			using (System.IO.StreamReader verRead = new System.IO.StreamReader(l2Tempver, Encoding.Default, true, 1))
			{
				sVersion = verRead.ReadLine();
				sFile = verRead.ReadLine();
				sUrlFile = verRead.ReadLine();

				CheckStatLabel.Text = "Version detected: " + sVersion;
				CheckStatLinkLabel.Text = sUrlFile;

				verRead.Close();
			}

			System.IO.File.Delete(l2Tempver);

			if (Application.ProductVersion == sVersion)
			{
				if (MessageBox.Show("You are already have latest release.\nWant to download again?",
					"Already up-date",
					MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.No)

					return;
			}
			else if (MessageBox.Show("Found new build! " + sVersion + "\nWant to download new?",
				"Found update",
				MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.No)
				return;

			if (await GetFileFromHttp(sUrlFile, sFile) == false)
			{
				_logger.Write(LogLevel.Information, "Downloading file is failed");

				MessageBox.Show("Can't download file from site.", "Can't download",
					MessageBoxButtons.OK, MessageBoxIcon.Error);
				return;
			}

			_logger.Write(LogLevel.Information, new string[]
			{
				"Downloaded is successfully:",
				"[" + sVersion + "][" + sFile + "]"
			});

			MessageBox.Show(
				"Download new update success. Last build is " + sVersion
				+ "\nNote: close tool, extract new build from file, your work folder",
				"Success.",
				MessageBoxButtons.OK, MessageBoxIcon.Information);
		}

		private void QuitButton_Click(object sender, EventArgs e)
		{
			this.Dispose();
		}

		private async Task<bool> GetFileFromHttp(string url, string file)
		{
			using (HttpClient httpClient = new HttpClient())
			{
				httpClient.DefaultRequestHeaders.CacheControl = new CacheControlHeaderValue
				{
					NoCache = true
				};

				_logger.Write(LogLevel.Trace, "Downloading [" + url + "]");

				var result = await httpClient.GetAsync(url);
				try
				{
					result.EnsureSuccessStatusCode();
				}
				catch (Exception ex)
				{
					_logger.Write(LogLevel.Error, new string[]
					{
						"Downloaded is failed:",
						"Remote file [" + url + "]",
						ex.Message
					});
					return false;
				}


				byte[] data = await result.Content.ReadAsByteArrayAsync();
				try
				{
					await System.IO.File.WriteAllBytesAsync(file, data);
					result.Content.Dispose();
				}
				catch (Exception ex)
				{
					_logger.Write(LogLevel.Error, new string[]
					{
						"Write file [" + file + "] error: ",
						ex.Message
					});

					MessageBox.Show(ex.Message);
					return false;
				}
			}

			return true;
		}

		private void SendMailButton_Click(object sender, EventArgs e)
		{
			System.Net.Mail.MailMessage feedback = new System.Net.Mail.MailMessage(MailFromTextBox.Text, SupportEmail, SubjectTextBox.Text,
				TextBox.Text);

			try
			{
				using (System.Net.Mail.SmtpClient mailclient = new System.Net.Mail.SmtpClient(MailSrvTextBox.Text))
				{
					mailclient.UseDefaultCredentials = true;
					mailclient.Send(feedback);
				}
			}
			catch (FormatException ex)
			{
				_logger.Write(LogLevel.Error, new string[]
				{
					"Send mail failed:",
					$"From [{MailFromTextBox.Text}]; To [{SupportEmail}]; Subject [{SubjectTextBox.Text}]",
					$"Text: [{TextBox.Text}]",
					ex.Message
				});

				MessageBox.Show(ex.Message);
			}
			catch (System.Net.Mail.SmtpException ex)
			{
				string exMessage = ex.Message
					+ (ex.InnerException == null ? String.Empty : $": " + ex.InnerException.Message);

				_logger.Write(LogLevel.Error, new string[]
				{
					"Send mail failed:",
					$"From [{MailFromTextBox.Text}]; To [{SupportEmail}]; Subject [{SubjectTextBox.Text}]",
					$"Text: [{TextBox.Text}]"
				});
				_logger.Write(LogLevel.Error, ex);

				MessageBox.Show(exMessage);
			}

			MessageBox.Show("Send complete.", "Success", MessageBoxButtons.OK, MessageBoxIcon.Information);
		}

		private void CheckStatLinkLabel_Click(object sender, EventArgs e)
		{
			Process.Start("iexplore.exe", CheckStatLinkLabel.Text);
		}
	}
}